recipe/meta.yaml

```
@@ -0,0 +1,120 @@
+{% set name = "ctransformers" %}
+{% set version = "0.2.27" %}
+{% set build_number = 2 %}
+
+{% set cuda_version = "12.4" %}
+
+package:
+  name: {{ name }}-suite
+  version: {{ version }}
+
+source:
+  url: https://github.com/marella/{{ name }}/archive/refs/tags/v{{ version }}.tar.gz
+  sha256: 62b556c0b1d355cdb681108268ee626e6ee6f39f4abc1429606583f348c55ee8
+  patches:
+    - patches/0001-Load-library-from-conda-prefix.patch
+    - patches/0002-Fix-install-path.patch
+    - patches/0003-Make-FindCUDAToolkit-required-when-building-for-GPU.patch
+
+build:
+  # Use a build number difference to ensure that the GPU
+  # variant is slightly preferred by conda's solver, so that it's preferentially
+  # installed where the platform supports it.
+  number: {{ build_number + 100 }}  # [ctransformers_variant == "gpu"]
+  number: {{ build_number }}        # [ctransformers_variant == "cpu"]
+  skip: true # [skip_cuda_prefect and (ctransformers_variant or "").startswith('gpu') and not (osx and arm64)]
+  # libcublas and cuda-libraries 12.4.* (a dependency of cuda-toolkit) isn't available on s390x
+  skip: true  # [s390x]
+
+
+outputs:
+  - name: ctransformers
+    script: build_base.sh  # [unix]
+    script: bld_base.bat   # [win]
+    build:
+      string: gpu_cuda{{ cuda_version | replace('.', '') }}py{{ CONDA_PY }}h{{PKG_HASH}}_{{ PKG_BUILDNUM }}  # [ctransformers_variant == "gpu" and (linux or win)]
+      string: gpu_mps_py{{ CONDA_PY }}h{{PKG_HASH}}_{{ PKG_BUILDNUM }}                                      # [ctransformers_variant == "gpu" and osx]
+      string: cpu_py{{ CONDA_PY }}h{{PKG_HASH}}_{{ PKG_BUILDNUM }}                                          # [ctransformers_variant == "cpu"]
+    requirements:
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        - {{ compiler('cuda') }}  # [ctransformers_variant == 'gpu' and not osx]
+        - cmake
+        - ninja
+        - patch     # [not win]
+        - m2-patch  # [win]
+      host:
+        - python
+        # setuptools build backend is used if CT_WHEEL=1
+        - setuptools
+        - wheel
+        - pip
+        # The CUDAToolkit has been set as required in 0003-Make-FindCUDAToolkit-required-when-building-for-GPU.patch
+        - cuda-toolkit {{ cuda_version }}*   # [ctransformers_variant == 'gpu' and not osx]
+      run:
+        - python
+        - huggingface_hub
+        - py-cpuinfo >=9.0.0,<10.0.0
+        - {{ pin_compatible('cuda-toolkit', max_pin='x.x') }}  # [ctransformers_variant == 'gpu' and not osx]
+        # __cuda isn't available on win
+        - __cuda >={{ cuda_version }}                          # [ctransformers_variant == 'gpu' and not (osx or win)]
```

```suggestion
        - __cuda >={{ cuda_version }}                          # [ctransformers_variant == 'gpu' and not (osx)]
```
This is defined on windows and should be enabled. Was you instance configured correctly?

---

I followed the instructions from the [rocket-platform](https://github.com/anaconda-distribution/rocket-platform/tree/main/machine-images#dev-instances):
Using remote desktop run these commands in sequence as Administrator:
& 'C:\prefect\install_cuda_driver.ps1'
& 'C:\prefect\install_cuda_12.3.0.ps1'

I wonder why there is a c`uda_12.3.0` and not `12.4.*`?

---

@cbouss If I remove `win` from the selector  as you suggested and remove `- {{ pin_compatible('cuda-version', max_pin='x.x') }}  # [(ctransformers_variant or "").startswith('gpu') and not osx]` then I got an error on the Win dev boxes:
```
TEST START: C:\Users\dev-admin\src\aggregate\cr\win-64\ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102.tar.bz2
WARNING: Multiple meta files found. The meta.yaml file in the base directory (C:\Users\DEV-AD~1\AppData\Local\Temp\2\tmpg3uw804o\info\recipe) will be used.
Adding in variants from C:\Users\DEV-AD~1\AppData\Local\Temp\2\tmpg3uw804o\info\recipe\conda_build_config.yaml
Renaming work directory 'C:\Users\dev-admin\src\aggregate\cr\ctransformers-suite_1736520222550\work' to 'C:\Users\dev-admin\src\aggregate\cr\ctransformers-suite_1736520222550\work_moved_ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102_win-64'
shutil.move(work)=C:\Users\dev-admin\src\aggregate\cr\ctransformers-suite_1736520222550\work, dest=C:\Users\dev-admin\src\aggregate\cr\ctransformers-suite_1736520222550\work_moved_ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102_win-64)
Reloading output folder: ...working... done
Solving environment (_test_env): ...working... failed
WARNING: failed to get package records, retrying.  exception was: Unsatisfiable dependencies for platform win-64: {MatchSpec("__cuda[version='>=12.4']"), MatchSpec("ctransformers==0.2.27=gpu_cuda124py312h1133c6f_102")}
Encountered problems while solving:
  - nothing provides __cuda >=12.4 needed by ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102

Could not solve for environment specs
The following package could not be installed
└─ ctransformers =0.2.27 gpu_cuda124py312h1133c6f_102 is not installable because it requires
   └─ __cuda >=12.4 *, which is missing on the system.
Reloading output folder: ...working... done
Solving environment (_test_env): ...working... failed
Reloading output folder: ...working... done
Solving environment (_test_env): ...working... failed
Reloading output folder: ...working... done
Solving environment (_test_env): ...working... failed
ERROR: Failed to get package records, max retries exceeded.
WARNING: Tests failed for ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102.tar.bz2 - moving package to C:\Users\dev-admin\src\aggregate\cr\broken
Traceback (most recent call last):
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 890, in get_install_actions
    _actions = _install_actions(prefix, index, specs, subdir=subdir)
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 1304, in install_actions
    txn = solver.solve_for_transaction(prune=False, ignore_pinned=False)
  File "C:\miniconda3\lib\site-packages\conda\core\solve.py", line 153, in solve_for_transaction
    unlink_precs, link_precs = self.solve_for_diff(
  File "C:\miniconda3\lib\site-packages\conda\core\solve.py", line 222, in solve_for_diff
    final_precs = self.solve_final_state(
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 181, in solve_final_state
    out_state = self._solving_loop(in_state, out_state, index)
  File "C:\miniconda3\lib\site-packages\conda\common\io.py", line 87, in decorated
    return f(*args, **kwds)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 307, in _solving_loop
    solved, outcome = self._solve_attempt(in_state, out_state, index, attempt=attempt)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 364, in _solve_attempt
    new_conflicts = self._maybe_raise_for_problems(outcome, index, out_state, old_conflicts)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 753, in _maybe_raise_for_problems
    self._maybe_raise_for_conda_build(
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 843, in _maybe_raise_for_conda_build
    raise exc
conda_libmamba_solver.conda_build_exceptions.ExplainedDependencyNeedsBuildingError: Unsatisfiable dependencies for platform win-64: {MatchSpec("__cuda[version='>=12.4']"), MatchSpec("ctransformers==0.2.27=gpu_cuda124py312h1133c6f_102")}
Encountered problems while solving:
  - nothing provides __cuda >=12.4 needed by ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102

Could not solve for environment specs
The following package could not be installed
└─ ctransformers =0.2.27 gpu_cuda124py312h1133c6f_102 is not installable because it requires
   └─ __cuda >=12.4 *, which is missing on the system.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 890, in get_install_actions
    _actions = _install_actions(prefix, index, specs, subdir=subdir)
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 1304, in install_actions
    txn = solver.solve_for_transaction(prune=False, ignore_pinned=False)
  File "C:\miniconda3\lib\site-packages\conda\core\solve.py", line 153, in solve_for_transaction
    unlink_precs, link_precs = self.solve_for_diff(
  File "C:\miniconda3\lib\site-packages\conda\core\solve.py", line 222, in solve_for_diff
    final_precs = self.solve_final_state(
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 181, in solve_final_state
    out_state = self._solving_loop(in_state, out_state, index)
  File "C:\miniconda3\lib\site-packages\conda\common\io.py", line 87, in decorated
    return f(*args, **kwds)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 307, in _solving_loop
    solved, outcome = self._solve_attempt(in_state, out_state, index, attempt=attempt)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 364, in _solve_attempt
    new_conflicts = self._maybe_raise_for_problems(outcome, index, out_state, old_conflicts)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 753, in _maybe_raise_for_problems
    self._maybe_raise_for_conda_build(
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 843, in _maybe_raise_for_conda_build
    raise exc
conda_libmamba_solver.conda_build_exceptions.ExplainedDependencyNeedsBuildingError: Unsatisfiable dependencies for platform win-64: {MatchSpec("__cuda[version='>=12.4']"), MatchSpec("ctransformers==0.2.27=gpu_cuda124py312h1133c6f_102")}
Encountered problems while solving:
  - nothing provides __cuda >=12.4 needed by ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102

Could not solve for environment specs
The following package could not be installed
└─ ctransformers =0.2.27 gpu_cuda124py312h1133c6f_102 is not installable because it requires
   └─ __cuda >=12.4 *, which is missing on the system.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 890, in get_install_actions
    _actions = _install_actions(prefix, index, specs, subdir=subdir)
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 1304, in install_actions
    txn = solver.solve_for_transaction(prune=False, ignore_pinned=False)
  File "C:\miniconda3\lib\site-packages\conda\core\solve.py", line 153, in solve_for_transaction
    unlink_precs, link_precs = self.solve_for_diff(
  File "C:\miniconda3\lib\site-packages\conda\core\solve.py", line 222, in solve_for_diff
    final_precs = self.solve_final_state(
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 181, in solve_final_state
    out_state = self._solving_loop(in_state, out_state, index)
  File "C:\miniconda3\lib\site-packages\conda\common\io.py", line 87, in decorated
    return f(*args, **kwds)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 307, in _solving_loop
    solved, outcome = self._solve_attempt(in_state, out_state, index, attempt=attempt)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 364, in _solve_attempt
    new_conflicts = self._maybe_raise_for_problems(outcome, index, out_state, old_conflicts)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 753, in _maybe_raise_for_problems
    self._maybe_raise_for_conda_build(
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 843, in _maybe_raise_for_conda_build
    raise exc
conda_libmamba_solver.conda_build_exceptions.ExplainedDependencyNeedsBuildingError: Unsatisfiable dependencies for platform win-64: {MatchSpec("__cuda[version='>=12.4']"), MatchSpec("ctransformers==0.2.27=gpu_cuda124py312h1133c6f_102")}
Encountered problems while solving:
  - nothing provides __cuda >=12.4 needed by ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102

Could not solve for environment specs
The following package could not be installed
└─ ctransformers =0.2.27 gpu_cuda124py312h1133c6f_102 is not installable because it requires
   └─ __cuda >=12.4 *, which is missing on the system.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\miniconda3\lib\site-packages\conda_build\build.py", line 3380, in test
    precs = environ.get_package_records(
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 937, in get_install_actions
    precs = get_package_records(
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 937, in get_install_actions
    precs = get_package_records(
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 937, in get_install_actions
    precs = get_package_records(
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 890, in get_install_actions
    _actions = _install_actions(prefix, index, specs, subdir=subdir)
  File "C:\miniconda3\lib\site-packages\conda_build\environ.py", line 1304, in install_actions
    txn = solver.solve_for_transaction(prune=False, ignore_pinned=False)
  File "C:\miniconda3\lib\site-packages\conda\core\solve.py", line 153, in solve_for_transaction
    unlink_precs, link_precs = self.solve_for_diff(
  File "C:\miniconda3\lib\site-packages\conda\core\solve.py", line 222, in solve_for_diff
    final_precs = self.solve_final_state(
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 181, in solve_final_state
    out_state = self._solving_loop(in_state, out_state, index)
  File "C:\miniconda3\lib\site-packages\conda\common\io.py", line 87, in decorated
    return f(*args, **kwds)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 307, in _solving_loop
    solved, outcome = self._solve_attempt(in_state, out_state, index, attempt=attempt)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 364, in _solve_attempt
    new_conflicts = self._maybe_raise_for_problems(outcome, index, out_state, old_conflicts)
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 753, in _maybe_raise_for_problems
    self._maybe_raise_for_conda_build(
  File "C:\miniconda3\lib\site-packages\conda_libmamba_solver\solver.py", line 843, in _maybe_raise_for_conda_build
    raise exc
conda_libmamba_solver.conda_build_exceptions.ExplainedDependencyNeedsBuildingError: Unsatisfiable dependencies for platform win-64: {MatchSpec("__cuda[version='>=12.4']"), MatchSpec("ctransformers==0.2.27=gpu_cuda124py312h1133c6f_102")}
Encountered problems while solving:
  - nothing provides __cuda >=12.4 needed by ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102

Could not solve for environment specs
The following package could not be installed
└─ ctransformers =0.2.27 gpu_cuda124py312h1133c6f_102 is not installable because it requires
   └─ __cuda >=12.4 *, which is missing on the system.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\miniconda3\Scripts\conda-build-script.py", line 10, in <module>
    sys.exit(execute())
  File "C:\miniconda3\lib\site-packages\conda_build\cli\main_build.py", line 648, in execute
    api.build(
  File "C:\miniconda3\lib\site-packages\conda_build\api.py", line 211, in build
    return build_tree(
  File "C:\miniconda3\lib\site-packages\conda_build\build.py", line 3668, in build_tree
    test(pkg, config=metadata.config.copy(), stats=stats)
  File "C:\miniconda3\lib\site-packages\conda_build\build.py", line 3405, in test
    tests_failed(
  File "C:\miniconda3\lib\site-packages\conda_build\build.py", line 3542, in tests_failed
    raise CondaBuildUserError("TESTS FAILED: " + os.path.basename(pkg))
conda_build.exceptions.CondaBuildUserError: TESTS FAILED: ctransformers-0.2.27-gpu_cuda124py312h1133c6f_102.tar.bz2
```

---

> I wonder why there is a cuda_12.3.0 and not 12.4.*?
It's historical - originally we worked on cuda 12.3 but by the time we really got started v12.4 was out so we released that instead.

---

Fixed